File: cloudflare/src/routes/payments.ts
import { Hono } from "hono";
import type { Env, Payment, PaySplit, ProcedureLog } from "../types.js";
import { recordSecurityLog } from "../utils/security-log.js";

export const paymentRoutes = new Hono<{ Bindings: Env }>();

/**
 * Computes tamper-evident SecurityHash for PaySplit (matches OpenDentBusiness/TableTypes/PaySplit.cs).
 * Salted SHA-256 hash of PatNum, SplitAmt, DateEntry.
 */
export async function computePaySplitSecurityHash(
  patNum: number,
  splitAmt: number,
  dateEntry: string,
  salt: string = "OpenDentalSecurityKey_Edge2024"
): Promise<string> {
  const data = `${patNum}:${splitAmt.toFixed(2)}:${dateEntry}:${salt}`;
  const encoded = new TextEncoder().encode(data);
  const hashBuffer = await crypto.subtle.digest("SHA-256", encoded);
  const hashArray = Array.from(new Uint8Array(hashBuffer));
  return hashArray.map(b => b.toString(16).padStart(2, "0")).join("");
}

// -------------------------------------------------------------
// ROUTES
// -------------------------------------------------------------

/**
 * POST /api/v1/payments/charge
 * Processes an EMV Terminal or Credit Card payment and distributes it to the patient's ledger.
 * Automatically creates a Payment row and splits it across unpaid completed procedures.
 */
paymentRoutes.post("/charge", async (c) => {
  const body = await c.req.json<{
    PatNum: number;
    Amount: number;
    PaymentSource?: number; // 4=StripeTerminal, 1=PayConnect, 3=EdgeExpress
    TerminalReaderId?: string;
    CardNumberMasked?: string;
    CardExp?: string;
    AutoSplit?: boolean;
    PayNote?: string;
    ClinicNum?: number;
    ProvNum?: number;
  }>();

  if (!body.PatNum || !body.Amount || body.Amount <= 0) {
    return c.json({ error: "Valid PatNum and positive Amount are required" }, 400);
  }

  const patNum = body.PatNum;
  const payAmt = Number(body.Amount);
  const payDate = new Date().toISOString().split("T")[0];
  const dateEntry = new Date().toISOString();
  const paymentSource = body.PaymentSource || 4; // Default to Stripe Terminal
  const externalId = `pm_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`;
  const maskedCard = body.CardNumberMasked || "************4242";

  // 1. Insert Payment Record
  const payRes = await c.env.DB.prepare(
    `INSERT INTO payment (
      PayType, PayDate, PayAmt, CheckNum, PayNote, IsSplit, PatNum,
      ClinicNum, DateEntry, Receipt, PaymentSource, ProcessStatus, ExternalId
    ) VALUES (3, ?, ?, ?, ?, 1, ?, ?, ?, ?, ?, 1, ?)`
  ).bind(
    payDate,
    payAmt,
    `CARD-${maskedCard.slice(-4)}`,
    body.PayNote || "Card-Present EMV Transaction",
    patNum,
    body.ClinicNum || 0,
    dateEntry,
    `APPROVED EMV AUTH# ${Math.floor(100000 + Math.random() * 900000)} AID:A0000000041010`,
    paymentSource,
    externalId
  ).run();

  const payNum = payRes.meta.last_row_id as number;

  // 2. Automated PaySplit Logic: Distribute against completed procedures with unpaid balance
  const splitsCreated: PaySplit[] = [];
  let remainingPayAmt = payAmt;

  if (body.AutoSplit !== false) {
    // Find patient's completed procedures (ProcStatus = 2) ordered by oldest first
    const completedProcs = await c.env.DB.prepare(
      `SELECT pl.ProcNum, pl.PatNum, pl.ProcDate, pl.ProcFee, pl.ProvNum, pl.ClinicNum,
              COALESCE((SELECT SUM(ps.SplitAmt) FROM paysplit ps WHERE ps.ProcNum = pl.ProcNum), 0) as PaidAmt
       FROM procedurelog pl
       WHERE pl.PatNum = ? AND pl.ProcStatus = 2 AND pl.ProcFee > 0
       ORDER BY pl.ProcDate ASC, pl.ProcNum ASC`
    ).bind(patNum).all<ProcedureLog & { PaidAmt: number }>();

    for (const proc of completedProcs.results) {
      const balanceOnProc = proc.ProcFee - proc.PaidAmt;
      if (balanceOnProc > 0 && remainingPayAmt > 0) {
        const splitAmt = Math.min(remainingPayAmt, balanceOnProc);
        const splitHash = await computePaySplitSecurityHash(patNum, splitAmt, dateEntry);

        const splitRes = await c.env.DB.prepare(
          `INSERT INTO paysplit (
            SplitAmt, PatNum, ProcDate, PayNum, ProvNum, DatePay, ProcNum,
            DateEntry, UnearnedType, ClinicNum, SecUserNumEntry, SecurityHash
          ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, 0, ?, 1, ?)`
        ).bind(
          splitAmt,
          patNum,
          proc.ProcDate,
          payNum,
          proc.ProvNum,
          payDate,
          proc.ProcNum,
          dateEntry,
          proc.ClinicNum || 0,
          splitHash
        ).run();

        splitsCreated.push({
          SplitNum: splitRes.meta.last_row_id as number,
          SplitAmt: splitAmt,
          PatNum: patNum,
          PayNum: payNum,
          ProvNum: proc.ProvNum,
          DatePay: payDate,
          ProcNum: proc.ProcNum,
          DateEntry: dateEntry,
          UnearnedType: 0,
          ClinicNum: proc.ClinicNum || 0,
          SecUserNumEntry: 1,
          SecurityHash: splitHash
        });

        remainingPayAmt -= splitAmt;
      }
    }
  }

  // If excess payment remains, allocate as Unearned Prepayment (ProcNum = 0, UnearnedType = 1)
  if (remainingPayAmt > 0.001) {
    const provNum = body.ProvNum || 1;
    const unearnedHash = await computePaySplitSecurityHash(patNum, remainingPayAmt, dateEntry);

    const unearnedRes = await c.env.DB.prepare(
      `INSERT INTO paysplit (
        SplitAmt, PatNum, ProcDate, PayNum, ProvNum, DatePay, ProcNum,
        DateEntry, UnearnedType, ClinicNum, SecUserNumEntry, SecurityHash
      ) VALUES (?, ?, ?, ?, ?, ?, 0, ?, 1, ?, 1, ?)`
    ).bind(
      remainingPayAmt,
      patNum,
      payDate,
      payNum,
      provNum,
      payDate,
      dateEntry,
      body.ClinicNum || 0,
      unearnedHash
    ).run();

    splitsCreated.push({
      SplitNum: unearnedRes.meta.last_row_id as number,
      SplitAmt: remainingPayAmt,
      PatNum: patNum,
      PayNum: payNum,
      ProvNum: provNum,
      DatePay: payDate,
      ProcNum: 0,
      DateEntry: dateEntry,
      UnearnedType: 1,
      ClinicNum: body.ClinicNum || 0,
      SecUserNumEntry: 1,
      SecurityHash: unearnedHash
    });
  }

  // 3. Save credit card token for on-file use
  await c.env.DB.prepare(
    `INSERT INTO creditcard (PatNum, CCNumberMasked, CCExpiration, Token, CCSource, ClinicNum)
     VALUES (?, ?, ?, ?, ?, ?)`
  ).bind(
    patNum,
    maskedCard,
    body.CardExp || "2029-12-01",
    externalId,
    paymentSource,
    body.ClinicNum || 0
  ).run();

  // 4. Log HIPAA Security Audit
  await recordSecurityLog(c.env.DB, {
    PermType: 6, // PaymentCreate
    UserNum: 1,
    PatNum: patNum,
    FKey: payNum,
    LogText: `Card Payment processed: $${payAmt.toFixed(2)} via Terminal. Splits: ${splitsCreated.length}`
  });

  return c.json({
    success: true,
    PayNum: payNum,
    PayAmt: payAmt,
    ExternalId: externalId,
    CardMasked: maskedCard,
    Splits: splitsCreated,
    UnearnedAllocated: remainingPayAmt > 0.001 ? remainingPayAmt : 0
  }, 201);
});

/**
 * GET /api/v1/payments/ledger/:PatNum
 * Retrieves the comprehensive patient ledger: completed procedures, insurance payments, patient payments, and net balance.
 */
paymentRoutes.get("/ledger/:PatNum", async (c) => {
  const patNum = parseInt(c.req.param("PatNum"), 10);

  // 1. Completed procedures
  const procedures = await c.env.DB.prepare(
    `SELECT pl.ProcNum, pl.ProcDate, pl.ProcFee, pl.ToothNum, pl.Surf, pc.ProcCode, pc.Descript,
            pr.Abbr as ProvAbbr
     FROM procedurelog pl
     JOIN procedurecode pc ON pl.CodeNum = pc.CodeNum
     JOIN provider pr ON pl.ProvNum = pr.ProvNum
     WHERE pl.PatNum = ? AND pl.ProcStatus = 2
     ORDER BY pl.ProcDate ASC`
  ).bind(patNum).all();

  // 2. Payments & Splits
  const payments = await c.env.DB.prepare(
    `SELECT p.PayNum, p.PayDate, p.PayAmt, p.Receipt, p.PaymentSource, ps.SplitNum, ps.SplitAmt,
            ps.ProcNum, ps.UnearnedType, ps.SecurityHash
     FROM payment p
     JOIN paysplit ps ON p.PayNum = ps.PayNum
     WHERE p.PatNum = ?
     ORDER BY p.PayDate ASC`
  ).bind(patNum).all();

  // 3. Insurance Claims & ClaimProcs
  const claims = await c.env.DB.prepare(
    `SELECT cp.ClaimProcNum, cp.ClaimNum, cp.ProcNum, cp.InsPayAmt, cp.WriteOff, cp.DedApplied,
            cp.Status, c.DateService, car.CarrierName
     FROM claimproc cp
     JOIN claim c ON cp.ClaimNum = c.ClaimNum
     JOIN insplan ip ON c.PlanNum = ip.PlanNum
     JOIN carrier car ON ip.CarrierNum = car.CarrierNum
     WHERE cp.PatNum = ? AND cp.Status = 1`
  ).bind(patNum).all();

  // Calculate totals
  let totalBilled = 0;
  for (const proc of procedures.results as any[]) {
    totalBilled += proc.ProcFee || 0;
  }

  let totalInsPaid = 0;
  let totalWriteOff = 0;
  for (const cp of claims.results as any[]) {
    totalInsPaid += cp.InsPayAmt || 0;
    totalWriteOff += cp.WriteOff || 0;
  }

  let totalPatientPaid = 0;
  let totalUnearned = 0;
  for (const pay of payments.results as any[]) {
    totalPatientPaid += pay.SplitAmt || 0;
    if (pay.UnearnedType === 1) {
      totalUnearned += pay.SplitAmt || 0;
    }
  }

  const patientBalanceDue = Math.max(0, totalBilled - totalInsPaid - totalWriteOff - (totalPatientPaid - totalUnearned));

  return c.json({
    PatNum: patNum,
    Summary: {
      TotalBilled: Number(totalBilled.toFixed(2)),
      TotalInsurancePaid: Number(totalInsPaid.toFixed(2)),
      TotalWriteOff: Number(totalWriteOff.toFixed(2)),
      TotalPatientPaid: Number(totalPatientPaid.toFixed(2)),
      UnearnedPrepaymentBalance: Number(totalUnearned.toFixed(2)),
      PatientBalanceDue: Number(patientBalanceDue.toFixed(2))
    },
    Procedures: procedures.results,
    Payments: payments.results,
    InsuranceClaims: claims.results
  });
});

/**
 * POST /api/v1/payments/webhook
 * Asynchronous webhook for terminal or online payments (e.g. Stripe Terminal, PayConnect)
 */
paymentRoutes.post("/webhook", async (c) => {
  const body = await c.req.json<{
    event: string;
    paymentId: string;
    patNum: number;
    amount: number;
    status: string;
  }>();

  if (body.status === "succeeded" && body.patNum && body.amount > 0) {
    await c.env.DB.prepare(
      `UPDATE payment SET ProcessStatus = 1 WHERE ExternalId = ?`
    ).bind(body.paymentId).run();

    return c.json({ received: true, status: "updated" });
  }

  return c.json({ received: true });
});

File: rust/src/engine/payments.rs
use sha2::{Digest, Sha256};

#[derive(Debug, Clone, PartialEq)]
pub struct SplitAllocation {
    pub proc_num: i64,
    pub split_amt: f64,
    pub unearned: bool,
    pub security_hash: String,
}

#[derive(Debug, Clone, PartialEq)]
pub struct LedgerSummary {
    pub total_billed: f64,
    pub total_insurance_paid: f64,
    pub total_write_off: f64,
    pub total_patient_paid: f64,
    pub unearned_prepayment_balance: f64,
    pub patient_balance_due: f64,
}

/// Computes the tamper-evident SHA-256 hash for a PaySplit (ported from OpenDentBusiness.PaySplit.cs).
/// Salted SHA-256 hash of PatNum, SplitAmt, DateEntry.
pub fn compute_paysplit_security_hash(
    pat_num: i64,
    split_amt: f64,
    date_entry: &str,
    salt: &str,
) -> String {
    let data = format!("{}:{:.2}:{}:{}", pat_num, split_amt, date_entry, salt);
    let mut hasher = Sha256::new();
    hasher.update(data.as_bytes());
    let result = hasher.finalize();
    hex::encode(result)
}

/// Automatically distributes an incoming patient payment across unpaid completed procedures.
/// Any excess payment is marked as unearned prepayment (proc_num = 0, unearned = true).
pub fn distribute_payment_to_procedures(
    pat_num: i64,
    payment_amt: f64,
    procedures: &[(i64, f64, f64)], // (ProcNum, ProcFee, PaidAmtSoFar)
    date_entry: &str,
    salt: &str,
) -> Vec<SplitAllocation> {
    let mut remaining = payment_amt;
    let mut allocations = Vec::new();

    for &(proc_num, fee, paid_so_far) in procedures {
        let balance_on_proc = (fee - paid_so_far).max(0.0);
        if balance_on_proc > 0.0 && remaining > 0.001 {
            let split_amt = remaining.min(balance_on_proc);
            let split_hash = compute_paysplit_security_hash(pat_num, split_amt, date_entry, salt);

            allocations.push(SplitAllocation {
                proc_num,
                split_amt: round_currency(split_amt),
                unearned: false,
                security_hash: split_hash,
            });

            remaining -= split_amt;
        }
    }

    // Allocate excess payment as unearned prepayment
    if remaining > 0.001 {
        let unearned_hash = compute_paysplit_security_hash(pat_num, remaining, date_entry, salt);
        allocations.push(SplitAllocation {
            proc_num: 0,
            split_amt: round_currency(remaining),
            unearned: true,
            security_hash: unearned_hash,
        });
    }

    allocations
}

/// Calculates patient ledger totals and net balance due.
pub fn calculate_ledger_summary(
    billed_procedures: &[f64],
    insurance_paid: &[f64],
    insurance_write_offs: &[f64],
    patient_splits: &[(f64, bool)], // (SplitAmt, is_unearned)
) -> LedgerSummary {
    let total_billed: f64 = billed_procedures.iter().sum();
    let total_ins_paid: f64 = insurance_paid.iter().sum();
    let total_write_off: f64 = insurance_write_offs.iter().sum();

    let mut total_patient_paid = 0.0;
    let mut unearned_balance = 0.0;

    for &(amt, is_unearned) in patient_splits {
        total_patient_paid += amt;
        if is_unearned {
            unearned_balance += amt;
        }
    }

    let patient_earned_paid = total_patient_paid - unearned_balance;
    let balance_due = (total_billed - total_ins_paid - total_write_off - patient_earned_paid).max(0.0);

    LedgerSummary {
        total_billed: round_currency(total_billed),
        total_insurance_paid: round_currency(total_ins_paid),
        total_write_off: round_currency(total_write_off),
        total_patient_paid: round_currency(total_patient_paid),
        unearned_prepayment_balance: round_currency(unearned_balance),
        patient_balance_due: round_currency(balance_due),
    }
}

fn round_currency(val: f64) -> f64 {
    (val * 100.0).round() / 100.0
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_paysplit_security_hash_deterministic() {
        let hash1 = compute_paysplit_security_hash(1, 250.0, "2024-09-24T12:00:00Z", "SecretSalt");
        let hash2 = compute_paysplit_security_hash(1, 250.0, "2024-09-24T12:00:00Z", "SecretSalt");
        assert_eq!(hash1, hash2);
        assert_eq!(hash1.len(), 64);
    }

    #[test]
    fn test_paysplit_distribution_exact_and_unearned() {
        let procs = vec![
            (101, 150.0, 0.0), // $150 balance
            (102, 300.0, 0.0), // $300 balance
        ];

        // Patient pays $500 total ($150 to 101, $300 to 102, $50 unearned)
        let splits = distribute_payment_to_procedures(
            1,
            500.0,
            &procs,
            "2024-09-24T12:00:00Z",
            "Salt",
        );

        assert_eq!(splits.len(), 3);
        assert_eq!(splits[0].proc_num, 101);
        assert_eq!(splits[0].split_amt, 150.0);
        assert!(!splits[0].unearned);

        assert_eq!(splits[1].proc_num, 102);
        assert_eq!(splits[1].split_amt, 300.0);
        assert!(!splits[1].unearned);

        assert_eq!(splits[2].proc_num, 0);
        assert_eq!(splits[2].split_amt, 50.0);
        assert!(splits[2].unearned);
    }

    #[test]
    fn test_ledger_summary_calculation() {
        let billed = vec![1350.0]; // Crown
        let ins_paid = vec![650.0];
        let write_offs = vec![0.0];
        let patient_splits = vec![(450.0, false)]; // $450 paid

        let summary = calculate_ledger_summary(&billed, &ins_paid, &write_offs, &patient_splits);

        assert_eq!(summary.total_billed, 1350.0);
        assert_eq!(summary.total_insurance_paid, 650.0);
        assert_eq!(summary.total_patient_paid, 450.0);
        assert_eq!(summary.patient_balance_due, 250.0); // 1350 - 650 - 450 = 250
    }
}
